util/kvencoder/kv_encoder.go

```diff
@@ -0,0 +1,111 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kvenc
+
+import (
+	"fmt"
+	"sync/atomic"
+
+	"github.com/pingcap/tidb/meta/autoid"
+	"github.com/pingcap/tidb/mysql"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb"
+	"github.com/pingcap/tidb/kv"
+	goctx "golang.org/x/net/context"
+)
+
+var _ KvEncoder = &kvEncoder{}
+var mockConnID uint64
+
+// KvPair is a key-value pair.
+type KvPair struct {
+	// Key is the key of the pair.
+	Key []byte
+	// Val is the value of the pair. if the op is delete, the len(Val) == 0
+	Val []byte
+}
+
+// KvEncoder is a encoder that transfer sql to key-value pairs.
+type KvEncoder interface {
+	// Encode transfer sql to kv pairs.
+	Encode(sql string) ([]KvPair, error)
+}
+
+type kvEncoder struct {
+	store kv.Storage
+	se    tidb.Session
+}
+
+// New new a KvEncoder
+func New(store kv.Storage, dbName string, idAlloc autoid.Allocator, skipConstraintCheck bool) (KvEncoder, error) {
+	kvEnc := &kvEncoder{}
+	err := kvEnc.initial(store, dbName, idAlloc)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	kvEnc.se.GetSessionVars().SkipConstraintCheck = skipConstraintCheck
+	return kvEnc, nil
+}
+
+func (e *kvEncoder) Encode(sql string) ([]KvPair, error) {
```

Comment:
I remember that we need to get a unique ID from the caller.

Reply:
caller can give an allocator, we will use it to alloc id.  so it can hook the alloc behavior.